mapiproxy/libmapistore/mapistore_python.c

```diff
@@ -2681,24 +2684,70 @@ static enum mapistore_error mapistore_python_message_set_read_flag(void *message
 	MAPISTORE_RETVAL_IF(strcmp("MessageObject", message->ob_type->tp_name),
 			    MAPISTORE_ERR_CONTEXT_FAILED, NULL);
 
-	/* Set PidTagMessageFlags */
-	pydict = PyDict_New();
-	if (pydict == NULL) {
-		DEBUG(0, ("[ERR][%s]: Unable to initialize Python Dictionary\n", __location__));
+	/* Retrieve the message dictionary and read PidTagMessageFlags */
+	proplist = PyList_New(0);
+	if (proplist == NULL) {
+		DEBUG(0, ("[ERR][%s][%s]: Unable to initialize Python list\n",
+			  pyobj->name, __location__));
+		PyErr_Print();
 		return MAPISTORE_ERR_CONTEXT_FAILED;
 	}
 
+	pres = PyObject_CallMethod(message, "get_properties", "O", proplist);
+	Py_DECREF(proplist);
+	if (pres == NULL) {
+		DEBUG(0, ("[ERR][%s][%s]: PyObject_CallMethod failed: \n",
+			  pyobj->name, __location__));
+		PyErr_Print();
+		return MAPISTORE_ERR_CONTEXT_FAILED;
+	}
+	if (PyDict_Check(pres) != true) {
+		DEBUG(0, ("[ERR][%s][%s]: dict expected to be returned but got '%s'\n",
+			  pyobj->name, __location__, pres->ob_type->tp_name));
+		Py_DECREF(pres);
+		return MAPISTORE_ERR_CONTEXT_FAILED;
+	}
 
 	key = PyString_FromString("PidTagMessageFlags");
-	val = PyLong_FromLong((uint32_t)flags);
-	if (PyDict_SetItem(pydict, key, val) == -1) {
-		DEBUG(0, ("[ERR][%s]: Unable to add entry to Python dictionary\n",
+	val = PyDict_GetItem(pres, key);
+	if (val == NULL) {
+		/* If the property is missing, set all flags to zero */
+		msg_flags = 0x0;
+	} else {
+		msg_flags = PyLong_AsUnsignedLong(val);
+	}
+	Py_DECREF(pres);
```

Comment:
`PyDict_GetItem()` returns a borrowed reference, so the reference count of `val` it isn't our responsibility.